### PR TITLE
Add code-lens to run pipeline

### LINKED ISSUE #587: [FEATURE] VSCode Extension: Run pipeline button
## User Story
1. As a {VSCode user}
2. I want {just to press a button to run a Jayvee pipeline}
3. So that {I don't need to do it via the command line myself}

## User Acceptance Criteria
- [ ] Pipelines are runnable individually by the interpreter
- [ ] The VSCode extension checks whether there is a correct version of the Jayvee CLI installed
- [ ] There is a button for users to press close to the pipeline definition
  - [ ] Clicking the button leads to an automated execution of the pipeline 

## Notes
CodeLenses might be a LSP native feature to implement this.

## Definitions of Done
- [ ] A PR has been opened and accepted
- [ ] All user acceptance criteria are met
- [ ] All tests are passing

### PR DESCRIPTION
**Pull request by georg-schwarz**
Closes #587.

Adds a `Run (once)` code lens over each pipeline that executes the jv CLI.

image

### CODE REVIEW

File: apps/vs-code-extension/src/commands/run-jv.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
```
Comment (rhazn) [inline apps/vs-code-extension/src/commands/run-jv.ts:1]: It's 2024 ;)

File: apps/vs-code-extension/src/commands/run-jv.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  assert(
+    isRunJayveeCodeLensPayload(payload),
+    `Payload ${JSON.stringify(
+      payload,
+    )} is not a valid RunJayveeCodeLensPayload`,
+  );
+  const filePath = payload.filePath;
+
+  const shell = window.createTerminal('Run Jayvee');
+  shell.sendText(`jv ${filePath} --pipeline ${payload.pipelineName}`);
```
Comment (rhazn) [inline apps/vs-code-extension/src/commands/run-jv.ts:34]: Would put a `-d` in there, no?

File: apps/interpreter/src/run-action.ts [hunk 3]
```diff
@@ -35,9 +34,9 @@ export async function runAction(
     return process.exit(ExitCode.FAILURE);
   }
 
-  const currentDir = getCurrentDir();
+  const currentDir = process.cwd();
   const workingDir = currentDir;
-  const absoluteFilePath = path.join(currentDir, filePath);
+  const filePathRelativeToCurrentDir = path.relative(currentDir, filePath);
```
Comment (rhazn) [inline apps/interpreter/src/run-action.ts:39]: Does this not introduce some weirdness again that makes calling a jv file dependent on where the interpreter is running?

File: libs/language-server/src/lib/lsp/jayvee-code-lens-provider.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  pipelineName: string;
+}
+
+export interface ShowPipeOutputCodeLensPayload {
+  pipeOutputName: string;
+  lineNumber: number;
+}
+
+export function isRunJayveeCodeLensPayload(
+  o: unknown,
```
Comment (rhazn) [inline libs/language-server/src/lib/lsp/jayvee-code-lens-provider.ts:23]: Naming could probably be better than `o`, no? At least something like 'payload'.

File: libs/language-server/src/lib/lsp/jayvee-code-lens-provider.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  return (
+    'filePath' in o &&
+    typeof o.filePath === 'string' &&
+    'pipelineName' in o &&
+    typeof o.pipelineName === 'string'
+  );
+}
+
+export function isShowPipeOutputCodeLensPayload(
+  o: unknown,
```
Comment (rhazn) [inline libs/language-server/src/lib/lsp/jayvee-code-lens-provider.ts:38]: Same here.

File: apps/vs-code-extension/src/commands/run-jv.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            `No matching jv versions: expected version ${requiredVersion} but found version ${installedVersion}`,
+          ),
+        );
+      }
+      return resolve();
+    });
+  });
+}
+
+function getExtensionVersion(context: ExtensionContext): string | undefined {
```
Comment (rhazn) [inline apps/vs-code-extension/src/commands/run-jv.ts:63]: This is really nice, could be used to other good effects like showing a warning if we ever add a versioning field to jv models.

**Review by rhazn — APPROVED** [review]
Not sure what the effects are of the directory logic changing in the `run-action.ts`, if you're confident in that change I'd trust your knowledge though.

Other comments are largely just style and can be ignored.

File: apps/interpreter/src/run-action.ts (hunk 3 above)
Comment (georg-schwarz) [reply apps/interpreter/src/run-action.ts:39]: > Not sure what the effects are of the directory logic changing in the run-action.ts, if you're confident in that change I'd trust your knowledge though.

Ah yes.. I noticed with our current version we stopped supporting absolute paths :(
The new directory logic enables absolute paths again, by mapping them to a relative path from the current dir.
I'm open to better solutions if you have some in mind :)

File: apps/vs-code-extension/src/commands/run-jv.ts (hunk 2 above)
Comment (georg-schwarz) [reply apps/vs-code-extension/src/commands/run-jv.ts:34]: Mhm, depends. We could also add a second code action `Debug (once)`?
But in general, we should offer an option to debug, I agree!

Comment (georg-schwarz) [reply apps/vs-code-extension/src/commands/run-jv.ts]: Added in [3a61993](https://github.com/jvalue/jayvee/pull/606/commits/3a61993b7ceb13680eec75d64aacca3ea2ea00b3)

File: apps/interpreter/src/run-action.ts (hunk 3 above)
Comment (rhazn) [reply apps/interpreter/src/run-action.ts:39]: I think at all times, all paths should be relative to the model location without exception :).

Comment (georg-schwarz) [reply apps/interpreter/src/run-action.ts:39]: Ah I mean for the interpreter, sorry!
So `jv /home/me/model.jv` wasn't working.
Import paths in models are always relative.

Comment (rhazn) [reply apps/interpreter/src/run-action.ts:39]: Ah, yeah, then I love it.

**Pull request merged**